### PR TITLE
New Anatomy of a Transaction

### PR DESCRIPTION
**Pull request by garious**
#### Problem

Transaction and Instruction API docs are trying to be both Rust docs and binary format docs, and not especially great at either.

#### Summary of Changes

* Let Rust docs be Rust docs: https://docs.rs/solana-sdk/0.22.3/solana_sdk/transaction/index.html
* Describe the binary format without any Rust idioms
* This PR revives the previous URL from "Anatomy of a Transaction" and deletes the two API URLs

### CODE REVIEW

File: book/src/transaction.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+### Message Header Format
+
+The message header contains three unsigned 8-bit values. The first value is a
+number of required signatures in the containing transaction. The second value
+is the number of read-only account addresses requiring signatures. These will
+be the first addresses in the message's vector of account addresses.  The third
+value in the header is the number of read-only account addresses not requiring
+signatures. These addresses immediately follow the addresses that require
+signatures. The remaining addresses are used to request write access to the
+cooresponding accounts.
```
Comment (CriesofCarrots) [inline book/src/transaction.md]: corresponding

File: book/src/transaction.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
+# Anatomy of a Transaction
+
+This chapter documents the binary format of a transaction.
+
+## Transaction Format
+
+A transaction contains a compact-array of signatures, followed by a message.
+The signatures are all signatures of the given message. The Solana runtime
+verifies that the number of signatures matches the number in the first 8 bits
+of the message header. It also verifies the signature was signed by the
```
Comment (CriesofCarrots) [inline book/src/transaction.md]:
```suggestion
of the message header. It also verifies that each signature was signed by the
```

File: book/src/transaction.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+private key corresponding to the public key at the same index in the message's
+account addresses vector.
```
Comment (CriesofCarrots) [inline book/src/transaction.md]: vector->array?

File: book/src/transaction.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+### Message Header Format
+
+The message header contains three unsigned 8-bit values. The first value is a
+number of required signatures in the containing transaction. The second value
+is the number of read-only account addresses requiring signatures. These will
+be the first addresses in the message's vector of account addresses.  The third
+value in the header is the number of read-only account addresses not requiring
+signatures. These addresses immediately follow the addresses that require
+signatures. The remaining addresses are used to request write access to the
```
Comment (CriesofCarrots) [inline book/src/transaction.md]: This section is ambiguous, and I think actually incorrect. The order of account addresses in the account_keys array is:
1. signed read-write keys
2. signed read-only keys
3. unsigned read-write keys
4. unsigned read-only keys.

Here a suggested rewording:
The message header contains three unsigned 8-bit values. The first value is the number of required signatures in the containing transaction. The second value is the number of those corresponding account addresses that are read-only. The addresses that require signatures appear at the beginning of the account address array, with addresses requesting write access first and read-only accounts following.

The third value in the message header is the number of read-only account addresses not requiring signatures. The addresses that do not require signatures follow the addresses that do in the account address array, with read-write accounts first and read-only accounts following.

---
Alternately, perhaps just use the ordered list in some fashion

File: book/src/transaction.md [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+This chapter documents the binary format of a transaction.
+
+## Transaction Format
+
+A transaction contains a [compact-array](#Compact-Array-Format) of signatures,
+followed by a [message](#Message-Format).  Each item in the signatures array is
+a [digital signature](#Signature-Format) of the given message. The Solana
+runtime verifies that the number of signatures matches the number in the first
+8 bits of the [message header](#Message-Header-Format). It also verifies the
+signature was signed by the private key corresponding to the public key at the
```
Comment (CriesofCarrots) [inline book/src/transaction.md]: Suggestion:
It also verifies that each signature was signed by the ...

File: book/src/transaction.md (hunk 2 above)
Comment (CriesofCarrots) [reply book/src/transaction.md]: Mm, apparently github did preserve these comments on outdated lines. Sorry for the dupe!

File: book/src/transaction.md (hunk 4 above)
Comment (t-nelson) [reply book/src/transaction.md]: > Alternately, perhaps just use the ordered list in some fashion
👍

File: book/src/transaction.md [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+value in the header is the number of read-only account addresses not requiring
+signatures. These addresses immediately follow the addresses that require
+signatures. The remaining addresses are used to request write access to the
+cooresponding accounts.
+
+### Blockhash Format
+
+A blockhash contains a 32-byte SHA-256 hash. It is used to indicate when a
+client last observed the ledger. Validators will reject transactions when the
+blockhash is too old.
```
Comment (t-nelson) [inline book/src/transaction.md]: Shameless plug... link to durable nonce docs? 😇

Comment (garious) [reply book/src/transaction.md]: maybe later. doesn't flow right, so not this PR

File: book/src/transaction.md (hunk 1 above)
Comment (garious) [reply book/src/transaction.md]: Fixed in 972d22e

File: book/src/transaction.md (hunk 3 above)
Comment (garious) [reply book/src/transaction.md]: Fixed in 972d22e

File: book/src/transaction.md (hunk 4 above)
Comment (garious) [reply book/src/transaction.md]: Fixed in 972d22e.  Added a section to describe the account addresses array

File: book/src/transaction.md [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+A compact-array is serialized as the array length, followed by each array item.
+The array length is a special multi-byte encoding called compact-u16.
+
+### Compact-u16 Format
+
+A compact-u16 is a multi-byte encoding of 16 bits. The first byte contains the
+lower 7 bits of the value in its lower 7 bits.  If the value is above 0x7f, the
+high bit is set and the next 7 bits of the value are placed into the lower 7
+bits of a second byte. If the value is above 0x3fff, the high bit is set and
+the remaining 2 bits of the value are placed into the lower 2 bits of the third
```
Comment (CriesofCarrots) [inline book/src/transaction.md]:
```suggestion
the remaining 2 bits of the value are placed into the lower 2 bits of a third
```
Indicates more clearly that the third byte does not exist if not needed

**Pull request merged**